File: AssetManager.ConvertedToC#/UserInterface/Forms/AssetManager/TrackDeviceForm.vb
Public Class TrackDeviceForm
    Private CurrentTrackingDevice As New DeviceObject
    Private CheckData As New DeviceTrackingObject

    Sub New(device As DeviceObject, parentForm As ExtendedForm)
        InitializeComponent()
        CurrentTrackingDevice = device
        Me.ParentForm = parentForm
        ClearAll()
        SetDates()
        SetGroups()
        GetCurrentTracking(CurrentTrackingDevice.GUID)
        LoadTracking()
        Show()
    End Sub

    Private Function GetCheckData() As Boolean
        If Not CurrentTrackingDevice.Tracking.IsCheckedOut Then
            Dim c As Control
            For Each c In CheckOutBox.Controls
                If TypeOf c Is TextBox Then
                    If c.Visible Then
                        If Trim(c.Text) = "" Then
                            Message("Please complete all fields.", vbOKOnly + vbExclamation, "Missing Data", Me)
                            Return False
                        End If
                    End If
                End If
            Next
        Else
            Dim c As Control
            For Each c In CheckInBox.Controls
                If TypeOf c Is TextBox Then
                    If Trim(c.Text) = "" Then
                        Message("Please complete all fields.", vbOKOnly + vbExclamation, "Missing Data", Me)
                        Return False
                    End If
                End If
            Next
        End If
        With CheckData
            .CheckoutTime = dtCheckOut.Value '.ToString(strDBDateTimeFormat)
            .DueBackTime = dtDueBack.Value '.ToString(strDBDateTimeFormat)
            .UseLocation = UCase(Trim(txtUseLocation.Text))
            .UseReason = UCase(Trim(txtUseReason.Text))
            .CheckinNotes = UCase(Trim(txtCheckInNotes.Text))
            .DeviceGUID = CurrentTrackingDevice.GUID
            .CheckoutUser = LocalDomainUser
            .CheckinTime = dtCheckIn.Value '.ToString(strDBDateTimeFormat)
            .CheckinUser = LocalDomainUser
        End With
        Return True
    End Function

    Private Sub GetCurrentTracking(strGUID As String)
        Using results As DataTable = DBFactory.GetDatabase.DataTableFromQueryString("SELECT * FROM " & TrackablesCols.TableName & " WHERE " & TrackablesCols.DeviceUID & "='" & strGUID & "' ORDER BY " & TrackablesCols.DateStamp & " DESC LIMIT 1") 'ds.Tables(0)
            If results.Rows.Count > 0 Then
                CurrentTrackingDevice.MapClassProperties(CurrentTrackingDevice, results)
            End If
        End Using
    End Sub

    Private Sub LoadTracking()
        txtAssetTag.Text = CurrentTrackingDevice.AssetTag
        txtDescription.Text = CurrentTrackingDevice.Description
        txtSerial.Text = CurrentTrackingDevice.Serial
        txtDeviceType.Text = GetDisplayValueFromCode(DeviceAttribute.EquipType, CurrentTrackingDevice.EquipmentType)
        If CurrentTrackingDevice.Tracking.IsCheckedOut Then
            dtCheckOut.Value = CurrentTrackingDevice.Tracking.CheckoutTime
            dtDueBack.Value = CurrentTrackingDevice.Tracking.DueBackTime
            txtUseLocation.Text = CurrentTrackingDevice.Tracking.UseLocation
            txtUseReason.Text = CurrentTrackingDevice.Tracking.UseReason
        End If
    End Sub

    Private Sub ClearAll()
        Dim c As Control
        For Each c In Me.Controls
            If TypeOf c Is GroupBox Then
                Dim gc As Control
                For Each gc In c.Controls
                    If TypeOf gc Is TextBox Then
                        Dim txt As TextBox = DirectCast(gc, TextBox)
                        txt.Text = ""
                    End If
                Next
            End If
        Next
    End Sub

    Private Sub SetDates()
        dtCheckOut.Value = Now
        dtCheckIn.Value = Now
        dtCheckOut.Enabled = False
        dtCheckIn.Enabled = False
    End Sub

    Private Sub SetGroups()
        CheckInBox.Enabled = CurrentTrackingDevice.Tracking.IsCheckedOut
        CheckOutBox.Enabled = Not CurrentTrackingDevice.Tracking.IsCheckedOut
    End Sub

    Private Sub CheckOut()
        Using trans = DBFactory.GetDatabase.StartTransaction, conn = trans.Connection
            Try
                If Not GetCheckData() Then Exit Sub
                OtherFunctions.SetWaitCursor(True, Me)
                Dim rows As Integer
                rows += DBFactory.GetDatabase.UpdateValue(DevicesCols.TableName, DevicesCols.CheckedOut, 1, DevicesCols.DeviceUID, CurrentTrackingDevice.GUID, trans)

                Dim CheckOutParams As New List(Of DBParameter)
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckType, CheckType.Checkout))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckoutTime, CheckData.CheckoutTime))
                CheckOutParams.Add(New DBParameter(TrackablesCols.DueBackDate, CheckData.DueBackTime))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckoutUser, CheckData.CheckoutUser))
                CheckOutParams.Add(New DBParameter(TrackablesCols.UseLocation, CheckData.UseLocation))
                CheckOutParams.Add(New DBParameter(TrackablesCols.Notes, CheckData.UseReason))
                CheckOutParams.Add(New DBParameter(TrackablesCols.DeviceUID, CheckData.DeviceGUID))
                rows += DBFactory.GetDatabase.InsertFromParameters(TrackablesCols.TableName, CheckOutParams, trans)

                If rows = 2 Then
                    trans.Commit()
                    Message("Device Checked Out!", vbOKOnly + vbInformation, "Success", Me)
                Else
                    trans.Rollback()
                    Message("Unsuccessful! The number of affected rows was not expected.", vbOKOnly + vbExclamation, "Unexpected Result", Me)
                End If
                ParentForm.RefreshData()
            Catch ex As Exception
                trans.Rollback()
                ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Finally
                Me.Dispose()
            End Try
        End Using
    End Sub

    Private Sub CheckIn()
        Using trans = DBFactory.GetDatabase.StartTransaction, conn = trans.Connection
            Try
                If Not GetCheckData() Then Exit Sub
                OtherFunctions.SetWaitCursor(True, Me)
                Dim rows As Integer
                rows += DBFactory.GetDatabase.UpdateValue(DevicesCols.TableName, DevicesCols.CheckedOut, 0, DevicesCols.DeviceUID, CurrentTrackingDevice.GUID, trans)

                Dim CheckOutParams As New List(Of DBParameter)
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckType, CheckType.Checkin))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckoutTime, CheckData.CheckoutTime))
                CheckOutParams.Add(New DBParameter(TrackablesCols.DueBackDate, CheckData.DueBackTime))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckinTime, CheckData.CheckinTime))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckoutUser, CheckData.CheckoutUser))
                CheckOutParams.Add(New DBParameter(TrackablesCols.CheckinUser, CheckData.CheckinUser))
                CheckOutParams.Add(New DBParameter(TrackablesCols.UseLocation, CheckData.UseLocation))
                CheckOutParams.Add(New DBParameter(TrackablesCols.Notes, CheckData.CheckinNotes))
                CheckOutParams.Add(New DBParameter(TrackablesCols.DeviceUID, CheckData.DeviceGUID))
                rows += DBFactory.GetDatabase.InsertFromParameters(TrackablesCols.TableName, CheckOutParams, trans)

                If rows = 2 Then
                    trans.Commit()
                    Message("Device Checked In!", vbOKOnly + vbInformation, "Success", Me)
                Else
                    trans.Rollback()
                    Message("Unsuccessful! The number of affected rows was not what was expected.", vbOKOnly + vbExclamation, "Unexpected Result", Me)
                End If
                ParentForm.RefreshData()
            Catch ex As Exception
                trans.Rollback()
                ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Finally
                Me.Dispose()
            End Try
        End Using
    End Sub

    Private Sub cmdCheckOut_Click(sender As Object, e As EventArgs) Handles cmdCheckOut.Click
        CheckOut()
    End Sub

    Private Sub cmdCheckIn_Click(sender As Object, e As EventArgs) Handles cmdCheckIn.Click
        CheckIn()
    End Sub

    Private Sub txtUseLocation_LostFocus(sender As Object, e As EventArgs) Handles txtUseLocation.LostFocus
        txtUseLocation.Text = UCase(Trim(txtUseLocation.Text))
    End Sub

    Private Sub txtUseReason_LostFocus(sender As Object, e As EventArgs) Handles txtUseReason.LostFocus
        txtUseReason.Text = UCase(Trim(txtUseReason.Text))
    End Sub

    Private Sub txtCheckInNotes_LostFocus(sender As Object, e As EventArgs) Handles txtCheckInNotes.LostFocus
        txtCheckInNotes.Text = UCase(Trim(txtCheckInNotes.Text))
    End Sub

End Class

File: AssetManager.ConvertedToC#/UserInterface/Forms/AssetManager/TrackDeviceForm.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using System.Data;
using Microsoft.VisualBasic;
using AssetManager.UserInterface.CustomControls;

namespace AssetManager.UserInterface.Forms.AssetManager
{
    public partial class TrackDeviceForm : ExtendedForm
    {
        private DeviceObject CurrentTrackingDevice = new DeviceObject();
        private DeviceTrackingObject CheckData = new DeviceTrackingObject();

        public TrackDeviceForm(DeviceObject device, ExtendedForm parentForm)
        {
            InitializeComponent();
            CurrentTrackingDevice = device;
            this.ParentForm = parentForm;
            ClearAll();
            SetDates();
            SetGroups();
            GetCurrentTracking(System.Convert.ToString(CurrentTrackingDevice.GUID));
            LoadTracking();
            Show();
        }

        private bool GetCheckData()
        {
            if (!CurrentTrackingDevice.Tracking.IsCheckedOut)
            {
                Control c = default(Control);
                foreach (Control tempLoopVar_c in CheckOutBox.Controls)
                {
                    c = tempLoopVar_c;
                    if (c is TextBox)
                    {
                        if (c.Visible)
                        {
                            if (Strings.Trim(System.Convert.ToString(c.Text)) == "")
                            {
                                OtherFunctions.Message("Please complete all fields.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Missing Data", this);
                                return false;
                            }
                        }
                    }
                }
            }
            else
            {
                Control c = default(Control);
                foreach (Control tempLoopVar_c in CheckInBox.Controls)
                {
                    c = tempLoopVar_c;
                    if (c is TextBox)
                    {
                        if (Strings.Trim(System.Convert.ToString(c.Text)) == "")
                        {
                            OtherFunctions.Message("Please complete all fields.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Missing Data", this);
                            return false;
                        }
                    }
                }
            }
            CheckData.CheckoutTime = dtCheckOut.Value; //.ToString(strDBDateTimeFormat)
            CheckData.DueBackTime = dtDueBack.Value; //.ToString(strDBDateTimeFormat)
            CheckData.UseLocation = Strings.Trim(System.Convert.ToString(txtUseLocation.Text)).ToUpper();
            CheckData.UseReason = Strings.Trim(System.Convert.ToString(txtUseReason.Text)).ToUpper();
            CheckData.CheckinNotes = Strings.Trim(System.Convert.ToString(txtCheckInNotes.Text)).ToUpper();
            CheckData.DeviceGUID = CurrentTrackingDevice.GUID;
            CheckData.CheckoutUser = GlobalConstants.LocalDomainUser;
            CheckData.CheckinTime = dtCheckIn.Value; //.ToString(strDBDateTimeFormat)
            CheckData.CheckinUser = GlobalConstants.LocalDomainUser;
            return true;
        }

        private void GetCurrentTracking(string strGUID)
        {
            using (DataTable results = DBFactory.GetDatabase().DataTableFromQueryString("SELECT * FROM " + TrackablesCols.TableName + " WHERE " + TrackablesCols.DeviceUID + "='" + strGUID + "' ORDER BY " + TrackablesCols.DateStamp + " DESC LIMIT 1")) //ds.Tables(0)
            {
                if (results.Rows.Count > 0)
                {
                    CurrentTrackingDevice.MapClassProperties(CurrentTrackingDevice, results);
                }
            }

        }

        private void LoadTracking()
        {
            txtAssetTag.Text = CurrentTrackingDevice.AssetTag;
            txtDescription.Text = CurrentTrackingDevice.Description;
            txtSerial.Text = CurrentTrackingDevice.Serial;
            txtDeviceType.Text = AttribIndexFunctions.GetDisplayValueFromCode(GlobalInstances.DeviceAttribute.EquipType, CurrentTrackingDevice.EquipmentType);
            if (CurrentTrackingDevice.Tracking.IsCheckedOut)
            {
                dtCheckOut.Value = CurrentTrackingDevice.Tracking.CheckoutTime;
                dtDueBack.Value = CurrentTrackingDevice.Tracking.DueBackTime;
                txtUseLocation.Text = CurrentTrackingDevice.Tracking.UseLocation;
                txtUseReason.Text = CurrentTrackingDevice.Tracking.UseReason;
            }
        }

        private void ClearAll()
        {
            Control c = default(Control);
            foreach (Control tempLoopVar_c in this.Controls)
            {
                c = tempLoopVar_c;
                if (c is GroupBox)
                {
                    Control gc = default(Control);
                    foreach (Control tempLoopVar_gc in c.Controls)
                    {
                        gc = tempLoopVar_gc;
                        if (gc is TextBox)
                        {
                            TextBox txt = (TextBox)gc;
                            txt.Text = "";
                        }
                    }
                }
            }
        }

        private void SetDates()
        {
            dtCheckOut.Value = DateTime.Now;
            dtCheckIn.Value = DateTime.Now;
            dtCheckOut.Enabled = false;
            dtCheckIn.Enabled = false;
        }

        private void SetGroups()
        {
            CheckInBox.Enabled = CurrentTrackingDevice.Tracking.IsCheckedOut;
            CheckOutBox.Enabled = !CurrentTrackingDevice.Tracking.IsCheckedOut;
        }

        private void CheckOut()
        {
            using (var trans = DBFactory.GetDatabase().StartTransaction())
            {
                using (var conn = trans.Connection)
                {
                    try
                    {
                        if (!GetCheckData())
                        {
                            return;
                        }
                        OtherFunctions.SetWaitCursor(true, this);
                        int rows = 0;
                        rows += System.Convert.ToInt32(DBFactory.GetDatabase().UpdateValue(DevicesCols.TableName, DevicesCols.CheckedOut, 1, DevicesCols.DeviceUID, CurrentTrackingDevice.GUID, trans));

                        List<DBParameter> CheckOutParams = new List<DBParameter>();
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckType, CheckType.Checkout));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckoutTime, CheckData.CheckoutTime));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.DueBackDate, CheckData.DueBackTime));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckoutUser, CheckData.CheckoutUser));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.UseLocation, CheckData.UseLocation));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.Notes, CheckData.UseReason));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.DeviceUID, CheckData.DeviceGUID));
                        rows += System.Convert.ToInt32(DBFactory.GetDatabase().InsertFromParameters(TrackablesCols.TableName, CheckOutParams, trans));

                        if (rows == 2)
                        {
                            trans.Commit();
                            OtherFunctions.Message("Device Checked Out!", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Information, "Success", this);
                        }
                        else
                        {
                            trans.Rollback();
                            OtherFunctions.Message("Unsuccessful! The number of affected rows was not expected.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Unexpected Result", this);
                        }
                        ParentForm.RefreshData();
                    }
                    catch (Exception ex)
                    {
                        trans.Rollback();
                        ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                    }
                    finally
                    {
                        this.Dispose();
                    }
                }
            }

        }

        private void CheckIn()
        {
            using (var trans = DBFactory.GetDatabase().StartTransaction())
            {
                using (var conn = trans.Connection)
                {
                    try
                    {
                        if (!GetCheckData())
                        {
                            return;
                        }
                        OtherFunctions.SetWaitCursor(true, this);
                        int rows = 0;
                        rows += System.Convert.ToInt32(DBFactory.GetDatabase().UpdateValue(DevicesCols.TableName, DevicesCols.CheckedOut, 0, DevicesCols.DeviceUID, CurrentTrackingDevice.GUID, trans));

                        List<DBParameter> CheckOutParams = new List<DBParameter>();
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckType, CheckType.Checkin));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckoutTime, CheckData.CheckoutTime));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.DueBackDate, CheckData.DueBackTime));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckinTime, CheckData.CheckinTime));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckoutUser, CheckData.CheckoutUser));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.CheckinUser, CheckData.CheckinUser));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.UseLocation, CheckData.UseLocation));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.Notes, CheckData.CheckinNotes));
                        CheckOutParams.Add(new DBParameter(TrackablesCols.DeviceUID, CheckData.DeviceGUID));
                        rows += System.Convert.ToInt32(DBFactory.GetDatabase().InsertFromParameters(TrackablesCols.TableName, CheckOutParams, trans));

                        if (rows == 2)
                        {
                            trans.Commit();
                            OtherFunctions.Message("Device Checked In!", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Information, "Success", this);
                        }
                        else
                        {
                            trans.Rollback();
                            OtherFunctions.Message("Unsuccessful! The number of affected rows was not what was expected.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Unexpected Result", this);
                        }
                        ParentForm.RefreshData();
                    }
                    catch (Exception ex)
                    {
                        trans.Rollback();
                        ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                    }
                    finally
                    {
                        this.Dispose();
                    }
                }
            }

        }

        private void cmdCheckOut_Click(object sender, EventArgs e)
        {
            CheckOut();
        }

        private void cmdCheckIn_Click(object sender, EventArgs e)
        {
            CheckIn();
        }

        private void txtUseLocation_LostFocus(object sender, EventArgs e)
        {
            txtUseLocation.Text = Strings.Trim(System.Convert.ToString(txtUseLocation.Text)).ToUpper();
        }

        private void txtUseReason_LostFocus(object sender, EventArgs e)
        {
            txtUseReason.Text = Strings.Trim(System.Convert.ToString(txtUseReason.Text)).ToUpper();
        }

        private void txtCheckInNotes_LostFocus(object sender, EventArgs e)
        {
            txtCheckInNotes.Text = Strings.Trim(System.Convert.ToString(txtCheckInNotes.Text)).ToUpper();
        }

    }
}